mava/systems/mat/anakin/mat.py

```diff
@@ -0,0 +1,597 @@
+    # computing the network entropy at train time.
+    params = actor_network.init(actor_net_key, init_x, init_action, jax.random.PRNGKey(0))
+    opt_state = actor_optim.init(params)
+
+    # Pack apply and update functions.
+    apply_fns = (
+        partial(actor_network.apply, method="get_actions"),
+        actor_network.apply,
+    )
+    # Get batched iterated update and replicate it to pmap it over cores.
+    learn = get_learner_fn(env, apply_fns, actor_optim.update, config)
+    learn = jax.pmap(learn, axis_name="device")
+
+    # Initialise environment states and timesteps: across devices and batches.
+    key, *env_keys = jax.random.split(
+        key, n_devices * config.system.update_batch_size * config.arch.num_envs + 1
+    )
+    env_states, timesteps = jax.vmap(env.reset, in_axes=(0))(
+        jnp.stack(env_keys),
+    )
+    reshape_states = lambda x: x.reshape(
+        (n_devices, config.system.update_batch_size, config.arch.num_envs) + x.shape[1:]
+    )
+    # (devices, update batch size, num_envs, ...)
+    env_states = tree.map(reshape_states, env_states)
+    timesteps = tree.map(reshape_states, timesteps)
+
+    # Load model from checkpoint if specified.
+    if config.logger.checkpointing.load_model:
+        loaded_checkpoint = Checkpointer(
+            model_name=config.logger.system_name,
+            **config.logger.checkpointing.load_args,  # Other checkpoint args
+        )
+        # Restore the learner state from the checkpoint
+        restored_params, _ = loaded_checkpoint.restore_params(input_params=params)
+        # Update the params
+        params = restored_params
+
+    # Define params to be replicated across devices and batches.
+    key, step_keys = jax.random.split(key)
+    replicate_learner = (params, opt_state, step_keys)
+
+    # Duplicate learner for update_batch_size.
+    broadcast = lambda x: jnp.broadcast_to(x, (config.system.update_batch_size, *x.shape))
+    replicate_learner = tree.map(broadcast, replicate_learner)
+
+    # Duplicate learner across devices.
+    replicate_learner = flax.jax_utils.replicate(replicate_learner, devices=jax.devices())
+    # Initialise learner state.
+    params, opt_state, step_keys = replicate_learner
+    init_learner_state = LearnerState(params, opt_state, step_keys, env_states, timesteps)
+
+    return learn, actor_network, init_learner_state
+
+
+def run_experiment(_config: DictConfig) -> float:
+    """Runs experiment."""
+    config = copy.deepcopy(_config)
+
+    n_devices = len(jax.devices())
+
+    # Create the enviroments for train and eval.
+    env, eval_env = environments.make(config)
+
+    # PRNG keys.
+    key, key_e, actor_net_key = jax.random.split(jax.random.PRNGKey(config.system.seed), num=3)
+
+    # Setup learner.
+    learn, actor_network, learner_state = learner_setup(env, (key, actor_net_key), config)
+
+    eval_keys = jax.random.split(key_e, n_devices)
+
+    def eval_act_fn(
```

Comment:
If we can here follow the ppo systems where we call a maker function from the evaluator instead of creating it here

Reply:
The reason PPO and SAC defines it in the evaluator is because the same function is used across many systems. Since MAT uses a different `act_fn` and is the only one that uses it it makes more sense to me to define it in the system run file.

Reply:
That's convincing, thanks for the explanation